### PR TITLE
Fix return for struct with methods

### LINKED ISSUE #585: Issue returning struct with implementations
While using the new feature added in the PR https://github.com/fzyzcjy/flutter_rust_bridge/pull/543, I found an issue while returning an struct containg an implementation.
Ex code:
```rs 
pub fn get_sum_struct() -> SumWith {
    SumWith {x: 21}
}

pub struct SumWith { pub x: u32 }

impl SumWith {
    pub fn sum(&self, y: u32) -> u32 { self.x + y }
}
```
> Highlight on the first function, `get_sum_struct`

If we now run the flutter_rust_bridge, no errors will be found. Anyhow, there is an issue in the `bridge_generated.dart` file.

Dart bridge generated code:
```dart
Future<SumWith> getSumStruct({dynamic hint}) => executeNormal(FlutterRustBridgeTask(
        callFfi: (port_) => inner.wire_get_sum(port_),
        parseSuccessData: _wire2api_sum_with,
        constMeta: kGetSumConstMeta,
        argValues: [],
        hint: hint,
      ));
```

The problem is with the `_wire2api_sum_with` var. It's type is of `SumWith Function(Native, dynamic)` instead of the expected `SumWith Function(dynamic)`.

I think the issue is quite simple to fix, and can be done by adding a nullable parameter, even thought I don't think they exist in rust. I haven't got the needed exprience in Rust to take hands in the matter and make a PR.

### PR DESCRIPTION
**Pull request by Zaitam**
Related to #587

Close #585 

Previous discussions: #585

## Checklist

- [x] New tests are added to ensure new features are working (probably by modifying the `frb_example/pure_dart` example).
- [ ] All existing and new tests are passing.
- [x] If this PR adds/changes features, documentations (in `book` folder) are modified as well.

### CODE REVIEW

**Comment by Zaitam** [comment]
@fzyzcjy If you can review the changes. They are basic (+6 lines, which aren't even lines) but I think this is what you meant to do

**Comment by fzyzcjy** [comment]
Good job. 

> - [ ] New tests are added to ensure new features are working (probably by modifying the frb_example/pure_dart example).

Seems that this task has not been finished :)

**Comment by fzyzcjy** [comment]
CI fails 👀 

Maybe you need to run the code generator locally and commit changes here

**Comment by lattice0** [comment]
Thanks for the help Zaitam!

**Comment by Zaitam** [comment]
Yeah sorry I took a few days off and can't fix it atm. I'll do it as soon as I can

**Comment by fzyzcjy** [comment]
Take your time!

File: frb_example/pure_dart/dart/lib/main.dart [hunk 1]
```diff
@@ -408,6 +408,13 @@ void main(List<String> args) async {
     expect(sum, equals(21 + 1 + 5));
   });
 
+  test('return SumWith array test', () async {
+    final List<SumWith> sumWithList = await api.getSumArray(a: 12, b: 23, c: 67);
+    expect(sumWithList[0].sum(y: 23, z: 67), 12 + 23 + 67);
+    expect(sumWithList[0].sum(y: 12, z: 67), 12 + 23 + 67);
+    expect(sumWithList[0].sum(y: 12, z: 23), 12 + 23 + 67);
```
Comment (fzyzcjy) [inline frb_example/pure_dart/dart/lib/main.dart:415]: nit: I guess you mean [0], [1], [2]?

**Review by fzyzcjy — COMMENTED** [review]
LGTM, waiting for CI to pass

File: frb_example/pure_dart/dart/lib/main.dart (hunk 1 above)
Comment (Zaitam) [reply frb_example/pure_dart/dart/lib/main.dart:415]: Yep...

**Comment by Zaitam** [comment]
@fzyzcjy, so I fixed the merge conflicts. I think the changes are done. Sorry if it took me long to update it, had other stuff. I skipped the 'add changes to book' bc this is already explained (Not this particular use case) But in vectors and arrays

**Comment by fzyzcjy** [comment]
Good job! I will review it in a minute.

**Comment by fzyzcjy** [comment]
@all-contributors please add @Zaitam for code

File: Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+[workspace]
```
Comment (fzyzcjy) [inline Cargo.toml:1]: uber-nit: maybe we can remove this cargo.toml here. If a workspace is needed (looks yes!), feel free to introduce it in a separate PR!

File: frb_example/pure_dart/dart/lib/main.dart (hunk 1 above)
Comment (fzyzcjy) [reply frb_example/pure_dart/dart/lib/main.dart:437]: Hey it is still `[0]` not [0] [1] [2]...

**Review by fzyzcjy — COMMENTED** [review]
Going to merge it, only two nits :)

File: Cargo.toml (hunk 2 above)
Comment (Zaitam) [reply Cargo.toml]: Yeah, deleted it...

File: frb_example/pure_dart/dart/lib/main.dart (hunk 1 above)
Comment (Zaitam) [reply frb_example/pure_dart/dart/lib/main.dart]: Sorry, IDK how I missed it again. I think after trying to merge the master, deleted the commit or un committed code...

**Comment by Zaitam** [comment]
Hey so I don't get the errors for the:
- macos-11, osx
- windows-2019
- ubuntu-20.04
- Valgrind test

The ones of the code linting are clear and easy. If u could explain any of them so I can fix it

**Comment by fzyzcjy** [comment]
for "run codegen" -- run code generator locally, and commit them.

for valgrind, check its log output

**Comment by Zaitam** [comment]
Under which dir?

**Comment by Zaitam** [comment]
So... The 3 issues (macos-11, windows-2019, ubuntu-20.04) are bc I need to run all of these folders test's locally?

**Comment by fzyzcjy** [comment]
Look at its log (e.g. https://github.com/fzyzcjy/flutter_rust_bridge/runs/8104588993?check_suite_focus=true) to see what is happening :)

**Comment by fzyzcjy** [comment]
Wait a bit... https://github.com/fzyzcjy/flutter_rust_bridge/runs/8107781203?check_suite_focus=true CI is failing even for master branch, which was working well!

* https://github.com/fzyzcjy/flutter_rust_bridge/issues/655

So, no worries about the CI before #655 is fixed :)

Sadly I cannot merge this PR until #655 is fixed, because as is widely recognized, we should merge when CI goes green.

**Comment by fzyzcjy** [comment]
👀 What happened @Zaitam

**Comment by Zaitam** [comment]
I was too behind in the commits so decided to delete the fork and start from 0. There were too many edits to the `src/generator/dart/mod.rs` that I got lost just by looking at it... Btw, how can I use the `just` cli to test the bridge?

**Comment by fzyzcjy** [comment]
I see, totally agree :)

> Btw, how can I use the just cli to test the bridge?

Run codegen? `just gen-bridge `

**Comment by fzyzcjy** [comment]
`just refresh_all` may also help

**Comment by Zaitam** [comment]
> error[E0658]: use of unstable library feature 'bool_to_option'
it is in .then_some
any ideas? (Clean copy of latest master version)

**Comment by fzyzcjy** [comment]
https://github.com/fzyzcjy/flutter_rust_bridge/commit/bdb7f11b5c8a5487355daac078d1166f657558f3

/cc @Desdaemon maybe your recent PR introduces non-stable feature?

**Comment by Desdaemon** [comment]
I think it was the recently stabilized `then_some` method in 1.64.

**Comment by Zaitam** [comment]
@fzyzcjy I didn't have much time to do it in the last days. I do confirm it was a rust problem, but I am also getting some flutter issues bc of I am unable to downgrade... Anyhow, i've been looking at the `ty_something.rs` files and say that the `self` class contained `ir`, `context`, etc. But couldn't find how to use `ty.dart_param_type()` similar to the use in the `generate_wire2api_func` function. Is it included in the self?

**Comment by fzyzcjy** [comment]
@Zaitam Maybe dig a bit into the source code and mimic other usages? This may help as well: http://cjycode.com/flutter_rust_bridge/contributing/design.html

**Pull request closed without merging**